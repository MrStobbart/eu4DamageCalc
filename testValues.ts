import { Player, BaseValues, UnitType, Unit, TechLevel, PlayerModifiers, Regiment } from "./interfaces";

export const techLevels: TechLevel[] = [
  { tech: 0, baseDamage: { infantry: { fire: 0.25, shock: 0.2 }, cavalry: { fire: 0.0, shock: 0.8 }, artillery: { fire: 0.0, shock: 0.0 } }, baseMoral: 2.0, tactics: 0.5, improvedFlanking: 0, combatWidth: 15, supplyLimit: 0 },
  { tech: 1, baseDamage: { infantry: { fire: 0.35, shock: 0.3 }, cavalry: { fire: 0.0, shock: 0.8 }, artillery: { fire: 0.0, shock: 0.0 } }, baseMoral: 2.0, tactics: 0.5, improvedFlanking: 0, combatWidth: 15, supplyLimit: 0 },
  { tech: 2, baseDamage: { infantry: { fire: 0.35, shock: 0.5 }, cavalry: { fire: 0.0, shock: 1.0 }, artillery: { fire: 0.0, shock: 0.0 } }, baseMoral: 2.0, tactics: 0.5, improvedFlanking: 0, combatWidth: 20, supplyLimit: 0 },
  { tech: 3, baseDamage: { infantry: { fire: 0.35, shock: 0.5 }, cavalry: { fire: 0.0, shock: 1.0 }, artillery: { fire: 0.0, shock: 0.0 } }, baseMoral: 2.5, tactics: 0.5, improvedFlanking: 0, combatWidth: 20, supplyLimit: 0 },
  { tech: 4, baseDamage: { infantry: { fire: 0.35, shock: 0.5 }, cavalry: { fire: 0.0, shock: 1.0 }, artillery: { fire: 0.0, shock: 0.0 } }, baseMoral: 3.0, tactics: 0.75, improvedFlanking: 0, combatWidth: 20, supplyLimit: 0 },
  { tech: 5, baseDamage: { infantry: { fire: 0.35, shock: 0.65 }, cavalry: { fire: 0.0, shock: 1.2 }, artillery: { fire: 0.0, shock: 0.0 } }, baseMoral: 3.0, tactics: 0.75, improvedFlanking: 0, combatWidth: 22, supplyLimit: 0.5 },
  { tech: 6, baseDamage: { infantry: { fire: 0.55, shock: 0.95 }, cavalry: { fire: 0.0, shock: 1.2 }, artillery: { fire: 0.0, shock: 0.0 } }, baseMoral: 3.0, tactics: 1.0, improvedFlanking: 0, combatWidth: 24, supplyLimit: 0.5 },
  { tech: 7, baseDamage: { infantry: { fire: 0.55, shock: 0.95 }, cavalry: { fire: 0.0, shock: 1.2 }, artillery: { fire: 1.0, shock: 0.05 } }, baseMoral: 3.0, tactics: 1.25, improvedFlanking: 0, combatWidth: 24, supplyLimit: 0.5 },
  { tech: 8, baseDamage: { infantry: { fire: 0.8, shock: 0.95 }, cavalry: { fire: 0.0, shock: 2.0 }, artillery: { fire: 1.0, shock: 0.05 } }, baseMoral: 3.0, tactics: 1.25, improvedFlanking: 0, combatWidth: 24, supplyLimit: 0.5 },
  { tech: 9, baseDamage: { infantry: { fire: 0.8, shock: 0.95 }, cavalry: { fire: 0.0, shock: 2.0 }, artillery: { fire: 1.0, shock: 0.05 } }, baseMoral: 3.0, tactics: 1.5, improvedFlanking: 0, combatWidth: 25, supplyLimit: 0.5 },
  { tech: 10, baseDamage: { infantry: { fire: 0.8, shock: 0.95 }, cavalry: { fire: 0.0, shock: 2.0 }, artillery: { fire: 1.0, shock: 0.05 } }, baseMoral: 3.0, tactics: 1.5, improvedFlanking: 0.25, combatWidth: 25, supplyLimit: 0.5 },
  { tech: 11, baseDamage: { infantry: { fire: 0.8, shock: 1.15 }, cavalry: { fire: 0.5, shock: 2.0 }, artillery: { fire: 1.0, shock: 0.05 } }, baseMoral: 3.0, tactics: 1.5, improvedFlanking: 0.25, combatWidth: 27, supplyLimit: 0.5 },
  { tech: 12, baseDamage: { infantry: { fire: 0.8, shock: 1.15 }, cavalry: { fire: 0.5, shock: 2.0 }, artillery: { fire: 1.0, shock: 0.05 } }, baseMoral: 3.0, tactics: 1.75, improvedFlanking: 0.25, combatWidth: 27, supplyLimit: 1 },
  { tech: 13, baseDamage: { infantry: { fire: 0.8, shock: 1.15 }, cavalry: { fire: 0.5, shock: 2.0 }, artillery: { fire: 1.4, shock: 0.15 } }, baseMoral: 3.0, tactics: 1.75, improvedFlanking: 0.25, combatWidth: 27, supplyLimit: 1 },
  { tech: 14, baseDamage: { infantry: { fire: 1.1, shock: 1.15 }, cavalry: { fire: 0.5, shock: 2.0 }, artillery: { fire: 1.4, shock: 0.15 } }, baseMoral: 3.0, tactics: 1.75, improvedFlanking: 0.25, combatWidth: 29, supplyLimit: 1 },
  { tech: 15, baseDamage: { infantry: { fire: 1.1, shock: 1.15 }, cavalry: { fire: 0.5, shock: 2.0 }, artillery: { fire: 1.4, shock: 0.15 } }, baseMoral: 4.0, tactics: 2.0, improvedFlanking: 0.25, combatWidth: 29, supplyLimit: 1 },
  { tech: 16, baseDamage: { infantry: { fire: 1.1, shock: 1.15 }, cavalry: { fire: 0.5, shock: 2.0 }, artillery: { fire: 2.4, shock: 0.25 } }, baseMoral: 4.0, tactics: 2.0, improvedFlanking: 0.25, combatWidth: 30, supplyLimit: 1 },
  { tech: 17, baseDamage: { infantry: { fire: 1.1, shock: 1.15 }, cavalry: { fire: 0.5, shock: 3.0 }, artillery: { fire: 2.4, shock: 0.25 } }, baseMoral: 4.0, tactics: 2.0, improvedFlanking: 0.25, combatWidth: 30, supplyLimit: 1.5 },
  { tech: 18, baseDamage: { infantry: { fire: 1.1, shock: 1.15 }, cavalry: { fire: 0.5, shock: 3.0 }, artillery: { fire: 2.4, shock: 0.25 } }, baseMoral: 4.0, tactics: 2.0, improvedFlanking: 0.5, combatWidth: 32, supplyLimit: 1.5 },
  { tech: 19, baseDamage: { infantry: { fire: 1.1, shock: 1.15 }, cavalry: { fire: 0.5, shock: 3.0 }, artillery: { fire: 2.4, shock: 0.25 } }, baseMoral: 4.0, tactics: 2.25, improvedFlanking: 0.5, combatWidth: 32, supplyLimit: 2 },
  { tech: 20, baseDamage: { infantry: { fire: 1.6, shock: 1.15 }, cavalry: { fire: 0.5, shock: 3.0 }, artillery: { fire: 2.4, shock: 0.25 } }, baseMoral: 4.0, tactics: 2.25, improvedFlanking: 0.5, combatWidth: 34, supplyLimit: 2 },
  { tech: 21, baseDamage: { infantry: { fire: 1.6, shock: 1.65 }, cavalry: { fire: 0.5, shock: 3.0 }, artillery: { fire: 2.4, shock: 0.25 } }, baseMoral: 4.0, tactics: 2.5, improvedFlanking: 0.5, combatWidth: 34, supplyLimit: 2 },
  { tech: 22, baseDamage: { infantry: { fire: 1.6, shock: 1.65 }, cavalry: { fire: 1.0, shock: 3.0 }, artillery: { fire: 4.4, shock: 0.35 } }, baseMoral: 4.0, tactics: 2.5, improvedFlanking: 0.5, combatWidth: 36, supplyLimit: 2 },
  { tech: 23, baseDamage: { infantry: { fire: 1.6, shock: 1.65 }, cavalry: { fire: 1.0, shock: 4.0 }, artillery: { fire: 4.4, shock: 0.35 } }, baseMoral: 4.0, tactics: 2.75, improvedFlanking: 1, combatWidth: 36, supplyLimit: 2 },
  { tech: 24, baseDamage: { infantry: { fire: 1.6, shock: 1.65 }, cavalry: { fire: 1.0, shock: 4.0 }, artillery: { fire: 4.4, shock: 0.35 } }, baseMoral: 4.0, tactics: 3.0, improvedFlanking: 1, combatWidth: 38, supplyLimit: 2 },
  { tech: 25, baseDamage: { infantry: { fire: 1.6, shock: 1.65 }, cavalry: { fire: 1.0, shock: 4.0 }, artillery: { fire: 6.4, shock: 0.45 } }, baseMoral: 4.0, tactics: 3.0, improvedFlanking: 1, combatWidth: 38, supplyLimit: 2 },
  { tech: 26, baseDamage: { infantry: { fire: 1.6, shock: 1.65 }, cavalry: { fire: 1.0, shock: 4.0 }, artillery: { fire: 6.4, shock: 0.45 } }, baseMoral: 5.0, tactics: 3.0, improvedFlanking: 1, combatWidth: 40, supplyLimit: 2 },
  { tech: 27, baseDamage: { infantry: { fire: 2.1, shock: 1.65 }, cavalry: { fire: 1.0, shock: 4.0 }, artillery: { fire: 6.4, shock: 0.45 } }, baseMoral: 5.0, tactics: 3.0, improvedFlanking: 1, combatWidth: 40, supplyLimit: 2.5 },
  { tech: 28, baseDamage: { infantry: { fire: 2.1, shock: 2.15 }, cavalry: { fire: 1.0, shock: 4.0 }, artillery: { fire: 6.4, shock: 0.45 } }, baseMoral: 5.0, tactics: 3.0, improvedFlanking: 1.25, combatWidth: 40, supplyLimit: 2.5 },
  { tech: 29, baseDamage: { infantry: { fire: 2.1, shock: 2.15 }, cavalry: { fire: 1.0, shock: 4.0 }, artillery: { fire: 6.4, shock: 0.45 } }, baseMoral: 5.0, tactics: 3.0, improvedFlanking: 1.25, combatWidth: 40, supplyLimit: 2.5 },
  { tech: 30, baseDamage: { infantry: { fire: 2.1, shock: 2.15 }, cavalry: { fire: 1.0, shock: 4.0 }, artillery: { fire: 6.4, shock: 0.45 } }, baseMoral: 6.0, tactics: 3.25, improvedFlanking: 1.5, combatWidth: 40, supplyLimit: 2.5 },
  { tech: 31, baseDamage: { infantry: { fire: 3.1, shock: 2.15 }, cavalry: { fire: 1.0, shock: 5.0 }, artillery: { fire: 6.4, shock: 0.45 } }, baseMoral: 6.0, tactics: 3.25, improvedFlanking: 1.5, combatWidth: 40, supplyLimit: 3 },
  { tech: 32, baseDamage: { infantry: { fire: 3.1, shock: 2.15 }, cavalry: { fire: 1.0, shock: 5.0 }, artillery: { fire: 8.4, shock: 0.55 } }, baseMoral: 6.0, tactics: 3.5, improvedFlanking: 1.5, combatWidth: 40, supplyLimit: 3 },
];

export const westernInfrantryUnits: Unit[] = [
  { tech: 1, fire: { off: 0, def: 0 }, shock: { off: 1, def: 0 }, moral: { off: 1, def: 0 }, name: "Halberd Infantry" },
  { tech: 1, fire: { off: 0, def: 0 }, shock: { off: 0, def: 0 }, moral: { off: 1, def: 1 }, name: "Latin Medieval Infantry" },
  { tech: 5, fire: { off: 0, def: 0 }, shock: { off: 1, def: 0 }, moral: { off: 2, def: 0 }, name: "Galloglaigh Infantry" },
  { tech: 5, fire: { off: 0, def: 0 }, shock: { off: 1, def: 0 }, moral: { off: 1, def: 1 }, name: "Longbow" },
  { tech: 5, fire: { off: 0, def: 0 }, shock: { off: 0, def: 1 }, moral: { off: 1, def: 1 }, name: "Men at Arms" },
  { tech: 9, fire: { off: 0, def: 0 }, shock: { off: 2, def: 1 }, moral: { off: 1, def: 1 }, name: "Condotta Infantry" },
  { tech: 9, fire: { off: 0, def: 0 }, shock: { off: 1, def: 1 }, moral: { off: 1, def: 2 }, name: "Landsknechten Infantry" },
  { tech: 9, fire: { off: 0, def: 0 }, shock: { off: 2, def: 0 }, moral: { off: 2, def: 1 }, name: "Reformed Galloglaigh Infantry" },
  { tech: 12, fire: { off: 1, def: 2 }, shock: { off: 2, def: 1 }, moral: { off: 3, def: 1 }, name: "Free Shooter Infantry" },
  { tech: 12, fire: { off: 1, def: 2 }, shock: { off: 1, def: 2 }, moral: { off: 2, def: 2 }, name: "Tercio Infantry" },
  { tech: 15, fire: { off: 1, def: 2 }, shock: { off: 3, def: 1 }, moral: { off: 3, def: 2 }, name: "Charge Infantry" },
  { tech: 15, fire: { off: 2, def: 2 }, shock: { off: 2, def: 1 }, moral: { off: 3, def: 2 }, name: "Maurician Infantry" },
  { tech: 19, fire: { off: 3, def: 2 }, shock: { off: 3, def: 2 }, moral: { off: 3, def: 2 }, name: "Gustavian Infantry" },
  { tech: 19, fire: { off: 2, def: 2 }, shock: { off: 3, def: 2 }, moral: { off: 4, def: 2 }, name: "Highlanders Infantry" },
  { tech: 19, fire: { off: 2, def: 2 }, shock: { off: 2, def: 3 }, moral: { off: 3, def: 3 }, name: "Reformed Tercio" },
  { tech: 23, fire: { off: 3, def: 2 }, shock: { off: 3, def: 2 }, moral: { off: 3, def: 3 }, name: "Caroline Infantry" },
  { tech: 23, fire: { off: 2, def: 3 }, shock: { off: 2, def: 3 }, moral: { off: 3, def: 3 }, name: "Grenzer Infantry" },
  { tech: 23, fire: { off: 3, def: 3 }, shock: { off: 2, def: 2 }, moral: { off: 3, def: 3 }, name: "Line Infantry" },
  { tech: 26, fire: { off: 3, def: 3 }, shock: { off: 3, def: 3 }, moral: { off: 4, def: 4 }, name: "Blue Coat Infantry" },
  { tech: 26, fire: { off: 4, def: 3 }, shock: { off: 3, def: 3 }, moral: { off: 4, def: 3 }, name: "Frederickian Infantry" },
  { tech: 26, fire: { off: 3, def: 4 }, shock: { off: 3, def: 3 }, moral: { off: 4, def: 3 }, name: "Red Coat Infantry" },
  { tech: 26, fire: { off: 4, def: 3 }, shock: { off: 3, def: 3 }, moral: { off: 3, def: 4 }, name: "White Coat Infantry" },
  { tech: 28, fire: { off: 4, def: 3 }, shock: { off: 3, def: 3 }, moral: { off: 4, def: 4 }, name: "Impulse Infantry" },
  { tech: 28, fire: { off: 3, def: 4 }, shock: { off: 3, def: 3 }, moral: { off: 4, def: 4 }, name: "Square Infantry" },
  { tech: 30, fire: { off: 4, def: 4 }, shock: { off: 3, def: 3 }, moral: { off: 4, def: 4 }, name: "Drill Infantry" },
  { tech: 30, fire: { off: 4, def: 4 }, shock: { off: 4, def: 3 }, moral: { off: 3, def: 4 }, name: "Jaeger Infantry" },
  { tech: 30, fire: { off: 4, def: 3 }, shock: { off: 4, def: 3 }, moral: { off: 4, def: 4 }, name: "Mixed Order Infantry" },
  { tech: 30, fire: { off: 4, def: 4 }, shock: { off: 4, def: 3 }, moral: { off: 4, def: 3 }, name: "Napoleonic Square" },
];

export const getFirstUnitByTech = (tech: number, type: UnitType = "infantry"): Unit => {
  const availableUnits = westernInfrantryUnits.filter((unit) => unit.tech <= tech);
  const unit = availableUnits[availableUnits.length - 1];
  if (unit) {
    return unit;
  }
  throw new Error(`No unit for tech ${tech} found. Tech must be between 0 and 30`);
};

export const getFirstUnitByName = (unitName: string): Unit => {
  const unit = westernInfrantryUnits.find((unit) => unit.name === unitName);
  if (unit) {
    return unit;
  }
  throw new Error(`Unit ${unitName} not found`);
};
