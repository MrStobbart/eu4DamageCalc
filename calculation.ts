import { getFirstUnitByTech, techLevels } from "./testValues";
import { BattleResult, Phase, Player, PlayerModifiers, Regiment } from "./interfaces";

const averageRoll = 4.5;

export const calculateRelativePips = (currentPlayer: PlayerModifiers, enemyPlayer: PlayerModifiers, phase: Phase, isMoral: boolean) => {
  const attackPips = !isMoral ? currentPlayer.unit[phase].off : currentPlayer.unit.moral.off;
  const enemyDefensePips = !isMoral ? enemyPlayer.unit[phase].def : enemyPlayer.unit.moral.def;
  const relativeGeneralPips = currentPlayer.generalPips[phase] - enemyPlayer.generalPips[phase];

  return Math.max(0, relativeGeneralPips) + attackPips - enemyDefensePips;
};

export const calculateCasualitiesMultiplier = (regiment: Regiment, enemyPlayer: PlayerModifiers, day: number, phase: Phase) => {
  const tech = techLevels[regiment.tech];
  const enemyTactics = techLevels[enemyPlayer.tech].tactics * enemyPlayer.discipline;
  const combatAbility = 1 + regiment.unitCombatAbility[regiment.type];

  return (((regiment.troops / 1000) * tech.baseDamage[regiment.type][phase]) / enemyTactics) * combatAbility * regiment.discipline * (1 + day / 100);
};

export const calculateMoralCasualities = (casualityMultiplier: number, currentPlayer: Regiment, enemyPlayer: Regiment, phase: Phase) => {
  const relativePips = calculateRelativePips(currentPlayer, enemyPlayer, phase, true);
  const maxMoral = techLevels[currentPlayer.tech].baseMoral * (1 + currentPlayer.moralBonus);

  return ((15 + 5 * (averageRoll + relativePips)) * casualityMultiplier * maxMoral) / 540 + 0.03;
};

export const calculateStrengthCasualities = (casualityMultiplier: number, currentPlayer: Regiment, enemyPlayer: Regiment, phase: Phase) => {
  const relativePips = calculateRelativePips(currentPlayer, enemyPlayer, phase, false);
  return (15 + 5 * (averageRoll + relativePips)) * casualityMultiplier * (1 + currentPlayer.damangeModifiers[phase].dealt) * (1 + enemyPlayer.damangeModifiers[phase].received);
};

const getBattleResult = (regiment1: Regiment, regiment2: Regiment, logs = true): BattleResult => {
  const remaining = { regiment1Remaining: { troops: regiment1.troops, moral: regiment1.moral }, regiment2Remaining: { troops: regiment2.troops, moral: regiment2.moral } };

  if (regiment1.moral > regiment2.moral) {
    if (logs) console.log("+++++ Player1 wins +++++");
    return { winner: "regiment1", ...remaining };
  }
  if (regiment2.moral > regiment1.moral) {
    if (logs) console.log("+++++ Player2 wins +++++");
    return { winner: "regiment2", ...remaining };
  }
  if (logs) console.log("+++++ Draw +++++ ");
  return { winner: "draw", ...remaining };
};

let phaseCounter = 1;
export const battleDay = (player1: Regiment, player2: Regiment, day = 1, phase: Phase = "fire", logs = true): BattleResult => {
  if (logs) console.log({ day, phase, player1: { moral: player1.moral, toops: player1.troops }, player2: { moral: player2.moral, toops: player2.troops } });

  if (player1.moral <= 0 || player2.moral <= 0) {
    return getBattleResult(player1, player2, logs);
  }
  const casualitiesModifierPlayer1 = calculateCasualitiesMultiplier(player1, player2, day, phase);
  const casualitiesModifierPlayer2 = calculateCasualitiesMultiplier(player2, player1, day, phase);

  const damageFromPlayer1 = {
    casualitiesModifierPlayer1,
    moral: calculateMoralCasualities(casualitiesModifierPlayer1, player1, player2, phase),
    strength: calculateStrengthCasualities(casualitiesModifierPlayer1, player1, player2, phase),
  };

  const damageFromPlayer2 = {
    casualitiesModifierPlayer2,
    moral: calculateMoralCasualities(casualitiesModifierPlayer2, player2, player1, phase),
    strength: calculateStrengthCasualities(casualitiesModifierPlayer2, player2, player1, phase),
  };

  player1.moral = player1.moral - damageFromPlayer2.moral;
  player1.troops = player1.troops - damageFromPlayer2.strength;

  player2.moral = player2.moral - damageFromPlayer1.moral;
  player2.troops = player2.troops - damageFromPlayer1.strength;

  if (logs) console.log({ damageFromPlayer1, damageFromPlayer2 });

  const changePhase = phaseCounter === 3;
  let nextPhase = phase;
  phaseCounter++;

  if (changePhase) {
    nextPhase = phase === "fire" ? "shock" : "fire";
    phaseCounter = 1;
  }

  if (logs) console.log({ changePhase, nextPhase, phaseCounter });

  return battleDay(player1, player2, (day = day + 1), nextPhase, logs);
};

export const baseModifiers: PlayerModifiers = {
  tech: 10,
  generalPips: { shock: 0, fire: 0 },
  moralBonus: 0,
  discipline: 1,
  damangeModifiers: { fire: { received: 0, dealt: 0 }, shock: { received: 0, dealt: 0 } },
  unitCombatAbility: { infantry: 0, cavalry: 0, artillery: 0 },
  type: "infantry",
  unit: getFirstUnitByTech(10, "infantry"),
};

export const getRegimentFromPlayerModifiers = (customModifiers: Partial<PlayerModifiers>): Regiment => {
  const unit = customModifiers.unit ? customModifiers.unit : getFirstUnitByTech(customModifiers.tech || baseModifiers.tech);
  const playerModifiers = { ...baseModifiers, ...customModifiers, unit };
  return { ...playerModifiers, troops: 1000, moral: techLevels[playerModifiers.tech].baseMoral * (1 + playerModifiers.moralBonus) };
};
