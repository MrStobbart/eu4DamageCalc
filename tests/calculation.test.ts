import { getFirstUnitByName } from "../testValues";
import { baseModifiers, calculateCasualitiesMultiplier, calculateMoralCasualities, calculateRelativePips, calculateStrengthCasualities, getRegimentFromPlayerModifiers } from "../calculation";

describe("calculateRelativePips", () => {
  it("calculates the relative pips correctly", () => {
    expect(calculateRelativePips(baseModifiers, baseModifiers, "shock", true)).toBe(1);
    expect(calculateRelativePips(baseModifiers, baseModifiers, "shock", false)).toBe(2);
    expect(calculateRelativePips(baseModifiers, baseModifiers, "fire", false)).toBe(0);
  });

  it("calculates the relative general pips correctly", () => {
    const player1 = { ...baseModifiers, generalPips: { shock: 2, fire: 1 } };
    const player2 = { ...baseModifiers, generalPips: { shock: 0, fire: 3 } };

    expect(calculateRelativePips(player1, player2, "shock", false)).toBe(4);
    expect(calculateRelativePips(player1, player2, "fire", false)).toBe(0);
    expect(calculateRelativePips(player2, player1, "fire", false)).toBe(2);
    expect(calculateRelativePips(player2, player1, "shock", false)).toBe(2);
  });
});

describe("calculateCasualitiesMultiplier", () => {
  it("calculates the correct multiplier", () => {
    const player1 = getRegimentFromPlayerModifiers({});
    const player2 = getRegimentFromPlayerModifiers({});

    expect(calculateCasualitiesMultiplier(player1, player2, 1, "fire")).toBe(0.5386666666666666);
    expect(calculateCasualitiesMultiplier(player1, player2, 1, "shock")).toBe(0.6396666666666666);
  });
});

describe("calculateMoralCasualities", () => {
  it("calculates the correct multiplier", () => {
    const player1 = getRegimentFromPlayerModifiers({});
    const player2 = getRegimentFromPlayerModifiers({});
    const multiplier = calculateCasualitiesMultiplier(player1, player2, 1, "fire");

    expect(calculateMoralCasualities(multiplier, player1, player2, "fire")).toBe(0.15718518518518518);
  });
});

describe("calculateStrengthCasualities", () => {
  it("calculates the correct multiplier", () => {
    const player1 = getRegimentFromPlayerModifiers({});
    const player2 = getRegimentFromPlayerModifiers({});

    const multiplierFire = calculateCasualitiesMultiplier(player1, player2, 1, "fire");
    expect(calculateStrengthCasualities(multiplierFire, player1, player2, "fire")).toBe(20.2);

    const multiplierShock = calculateCasualitiesMultiplier(player1, player2, 1, "shock");
    expect(calculateStrengthCasualities(multiplierShock, player1, player2, "shock")).toBe(30.384166666666665);
  });
});

describe("getRegimentFromPlayerModifiers", () => {
  it("has the correct default regiment", () => {
    expect(getRegimentFromPlayerModifiers({})).toMatchSnapshot();
  });
  it("sets the unit according to tech", () => {
    const regiment = getRegimentFromPlayerModifiers({ tech: 20 });
    expect(regiment.unit.name).toBe("Reformed Tercio");
    expect(regiment).toMatchSnapshot();
  });
  it("accepts custom units", () => {
    expect(getRegimentFromPlayerModifiers({ tech: 25, unit: getFirstUnitByName("Reformed Tercio") })).toMatchSnapshot();
  });
});
