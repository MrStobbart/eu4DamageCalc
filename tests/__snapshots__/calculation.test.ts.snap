// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRegimentFromPlayerModifiers accepts custom units 1`] = `
Object {
  "currentMoral": 4,
  "damangeModifiers": Object {
    "fire": Object {
      "dealt": 0,
      "received": 0,
    },
    "shock": Object {
      "dealt": 0,
      "received": 0,
    },
  },
  "discipline": 1,
  "generalPips": Object {
    "fire": 0,
    "shock": 0,
  },
  "moralBonus": 0,
  "tech": 25,
  "troops": 1000,
  "type": "infantry",
  "unit": Object {
    "fire": Object {
      "def": 2,
      "off": 2,
    },
    "moral": Object {
      "def": 3,
      "off": 3,
    },
    "name": "Reformed Tercio",
    "shock": Object {
      "def": 3,
      "off": 2,
    },
    "tech": 19,
  },
  "unitCombatAbility": Object {
    "artillery": 0,
    "cavalry": 0,
    "infantry": 0,
  },
}
`;

exports[`getRegimentFromPlayerModifiers has the correct default regiment 1`] = `
Object {
  "currentMoral": 3,
  "damangeModifiers": Object {
    "fire": Object {
      "dealt": 0,
      "received": 0,
    },
    "shock": Object {
      "dealt": 0,
      "received": 0,
    },
  },
  "discipline": 1,
  "generalPips": Object {
    "fire": 0,
    "shock": 0,
  },
  "moralBonus": 0,
  "tech": 10,
  "troops": 1000,
  "type": "infantry",
  "unit": Object {
    "fire": Object {
      "def": 0,
      "off": 0,
    },
    "moral": Object {
      "def": 1,
      "off": 2,
    },
    "name": "Reformed Galloglaigh Infantry",
    "shock": Object {
      "def": 0,
      "off": 2,
    },
    "tech": 9,
  },
  "unitCombatAbility": Object {
    "artillery": 0,
    "cavalry": 0,
    "infantry": 0,
  },
}
`;

exports[`getRegimentFromPlayerModifiers sets the unit according to tech 1`] = `
Object {
  "currentMoral": 4,
  "damangeModifiers": Object {
    "fire": Object {
      "dealt": 0,
      "received": 0,
    },
    "shock": Object {
      "dealt": 0,
      "received": 0,
    },
  },
  "discipline": 1,
  "generalPips": Object {
    "fire": 0,
    "shock": 0,
  },
  "moralBonus": 0,
  "tech": 20,
  "troops": 1000,
  "type": "infantry",
  "unit": Object {
    "fire": Object {
      "def": 2,
      "off": 2,
    },
    "moral": Object {
      "def": 3,
      "off": 3,
    },
    "name": "Reformed Tercio",
    "shock": Object {
      "def": 3,
      "off": 2,
    },
    "tech": 19,
  },
  "unitCombatAbility": Object {
    "artillery": 0,
    "cavalry": 0,
    "infantry": 0,
  },
}
`;
